HCore-Web/Providers/Impl/DownloadProcessingProxyUrlProviderImpl.cs

```diff
@@ -0,0 +1,197 @@
+
+            UriBuilder proxyUri = new UriBuilder(proxyBaseUrl);
+            if (proxyUri.Query != null && proxyUri.Query.Length > 1)
+            {
+                proxyUri.Query = proxyUri.Query.Substring(1) + "&" + uriQueryPayload;
+            }
+            else
+            {
+                proxyUri.Query = uriQueryPayload;
+            }
+
+            proxyUri.Query = proxyUri.Query.Substring(1) + "&fileName=" + Uri.EscapeDataString(fileName)
+                + (downloadMimeType != null ? "&mime=" + Uri.EscapeDataString(downloadMimeType) : "")
+                + "&u=" + Uri.EscapeDataString(downloadSourceUri.ToString())
+            ;
+            return Task.FromResult(proxyUri.Uri);
+        }
+
+        public virtual async Task<IDownloadFileData> GetFileDataAsync(HttpRequest request, Stream inputData)
+        {
+            Uri downloadUri = new Uri(request.Query["u"]);
+            string fileName = request.Query["fileName"];
+            string mimeType = request.Query["mime"];
+            string characterSet = "utf-8";
+            string originalHash = ConvertParametersFromQueryData(request);
+
+            if (!IsHashValid(originalHash, downloadUri, fileName, mimeType))
+            {
+                throw new UnauthorizedAccessException("Invalid query data");
+            }
+
+            Stream fileData = inputData;
+            if (fileData == null)
+            {
+                var remoteRequest = new HttpRequestMessage(HttpMethod.Get, downloadUri);
+                var client = _serviceProvider.GetService<HttpClient>();
+                var response = await client.SendAsync(remoteRequest).ConfigureAwait(false);
+
+                if (response.IsSuccessStatusCode)
+                {
+                    fileData = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
+
+                    if (string.IsNullOrEmpty(mimeType))
+                    {
+                        mimeType = response.Content.Headers.ContentType.MediaType;
+                    }
+
+                    characterSet = response.Content.Headers.ContentType.CharSet;
+                }
+                else
+                {
+                    _logger.LogError(
+                        $"Failed to download file from source location {downloadUri}. " +
+                        $"file name: {JsonConvert.SerializeObject(fileName)}. " +
+                        $"HTTP result was: {response.StatusCode} {response.ReasonPhrase}, response body: {response.Content}"
+                    );
+
+                    throw new HttpRequestException("Failed to download file from source location!");
+                }
+            }
+
+            return new DownloadFileDataImpl()
+            {
+                data = fileData,
+                FileName = fileName,
+                CharacterSet = characterSet,
+                MimeType = mimeType,
+            };
+        }
+
+        public virtual Uri GetSourceUri(HttpRequest request)
```

Comment:
What is this used for?

Reply:
It can be used by the proxy controller ore more stacked instances in case anyone else want to download the original source and pre-process it. Then we are able to stack multiple processing, like additional processing of images by piping it through cloudinary.

So, we could create an additional proxy processor implementing the same interface and stack these. However, stacking is not yet properly finished, as time and test case is lacking. There is more work to do, as the proxy controller must know which processors to stack. I would prefer to stack them in the proxy code locally instead of stacking the request by fetching from another local controller by URL.